include/dice/template-library/pool_allocator.hpp

```diff
@@ -0,0 +1,205 @@
+#ifndef DICE_TEMPLATELIBRARY_POOLALLOCATOR_HPP
+#define DICE_TEMPLATELIBRARY_POOLALLOCATOR_HPP
+
+#include <boost/pool/pool.hpp>
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <type_traits>
+
+namespace dice::template_library {
+
+	/**
+	 * A memory pool or arena that is efficient for allocations which are smaller or equal in size
+	 * for one of `bucket_sizes...`.
+	 *
+	 * The implementation consists of one arena per provided bucket size.
+	 * An allocation will be placed into the first bucket where it can fit.
+	 * Allocations that do not fit into any bucket are fulfilled with calls to `new`.
+	 *
+	 * @tparam bucket_sizes allocation sizes (in bytes) for the underlying arenas
+	 */
+	template<size_t ...bucket_sizes>
+	struct pool;
+
+	/**
+	 * `std`-style allocator that allocates into an underlying pool.
+	 * The bucket size used for allocation is `sizeof(T) * n_elems`.
+	 *
+	 * @tparam T type to be allocated
+	 * @tparam bucket_sizes same as for `pool<bucket_sizes...>`
+	 */
+	template<typename T, size_t ...bucket_sizes>
+	struct pool_allocator {
+		using value_type = T;
+		using pointer = T *;
+		using const_pointer = T const *;
+		using void_pointer = void *;
+		using const_void_pointer = void const *;
+		using size_type = size_t;
+		using difference_type = std::ptrdiff_t;
+
+		using propagate_on_container_copy_assignment = std::true_type;
+		using propagate_on_container_move_assignment = std::true_type;
+		using propagate_on_container_swap = std::true_type;
+		using is_always_equal = std::false_type;
+
+    	template<typename U>
+    	struct rebind {
+    		using other = pool_allocator<U, bucket_sizes...>;
+    	};
+
+	private:
+		template<typename, size_t ...>
+		friend struct pool_allocator;
+
+    	pool<bucket_sizes...> *pool_;
+
+	public:
+		explicit constexpr pool_allocator(pool<bucket_sizes...> &parent_pool) noexcept
```

Comment:
Could the allocator create its own pool (instead of relying on one to be constructed)?

Reply:
It could use a `shared_ptr` but it cannot exclusively own the pool, we need multiple allocator instances (of different types, e.g. `pool_allocator<int>` `pool_allocator<long>`) to allocate in the same pool.

This is for std container purposes and our own usecase of allocating multiple different node types.

I was pondering if this makes sense but ultimately decided to use a similar design to what metall uses.

I'm still somewhat unsure as the `shared_ptr` design would indeed be more convenient to use.

Reply:
I was basing this design on metall, but I guess they are only doing it this way because they are working with persistent memory. So maybe the shared ptr design makes more sense afterall

Reply:
@bigerl opinion?

Pros of current design: best perf (although a refcount increase per copy is really not that expensive)
Cons of current design: easy UB source, inconvenient to use (you always need to take care to keep the pool alive)

Reply:
I would opt for the `shared_ptr` to keep it simple. Only, if we really see performance issues we should revert back to providing the pool as an constructor arguement.